src/Interpreters/TemporaryDataOnDisk.cpp

```diff
@@ -163,13 +218,34 @@ TemporaryFileStream::TemporaryFileStream(TemporaryFileOnDiskHolder file_, const
 {
 }
 
-void TemporaryFileStream::write(const Block & block)
+TemporaryFileStream::TemporaryFileStream(FileSegmentsHolder && segments_, const Block & header_, TemporaryDataOnDisk * parent_)
+    : parent(parent_)
+    , header(header_)
+    , segment_holder(std::move(segments_))
+{
+    if (segment_holder.file_segments.size() != 1)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "TemporaryFileStream can be created only from single segment");
+    auto & segment = segment_holder.file_segments.front();
+    auto out_buf = std::make_unique<WriteBufferToFileSegment>(segment->detachWriter(), segment.get());
```

Comment:
May be do `segment->detachWriter()` in constructor of `WriteBufferToFileSegment` as we anyway pass segment there?